In [15]:
import matplotlib.pyplot as plt
import pandas

import seaborn
seaborn.set_context('talk')
pandas.set_option("display.max_columns", None)

melb_df = pandas.read_csv('../data/non_nan_values_melb.csv')
melb_df[:3]


Unnamed: 0      Suburb          Address  Rooms Type      Price Method  \
0           0  Abbotsford     85 Turner St      2    h  1480000.0      S   
1           1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S   
2           2  Abbotsford     5 Charles St      3    h  1465000.0     SP   

  SellerG       Date  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0  Biggin  3/12/2016       2.5    3067.0       2.0       1.0  1.0     202.0   
1  Biggin  4/02/2016       2.5    3067.0       2.0       1.0  0.0     156.0   
2  Biggin  4/03/2017       2.5    3067.0       3.0       2.0  0.0     134.0   

   BuildingArea CouncilArea  Lattitude  Longtitude             Regionname  \
0    137.374429       Yarra   -37.7996    144.9984  Northern Metropolitan   
1     79.000000       Yarra   -37.8079    144.9934  Northern Metropolitan   
2    150.000000       Yarra   -37.8093    144.9944  Northern Metropolitan   

   Propertycount  suburb_count CouncilArea_mode  
0         4019.0            56            Yarra  
1         4019.0            56            Yarra  
2         4019.0            56            Yarra

In [16]:
categorical_vars = ['Suburb', 'Type', 'Method', 'SellerG', 'CouncilArea', 'Regionname']
numerical_vars = ["Bedroom2", "BuildingArea", "Longtitude", "Car", "Postcode", "Bathroom", "Distance", "Lattitude", "Rooms", "suburb_count", "Price", "Propertycount", "Landsize"]

In [44]:
categorical_vars

['Suburb', 'Type', 'Method', 'SellerG', 'CouncilArea', 'Regionname']

In [45]:
numerical_vars

['Car',
 'Distance',
 'Landsize',
 'BuildingArea',
 'Bedroom2',
 'Bathroom',
 'Date',
 'Rooms',
 'Price',
 'Postcode',
 'Lattitude',
 'suburb_count',
 'Propertycount',
 'CouncilArea_mode',
 'Unnamed: 0',
 'Address',
 'Longtitude']

In [18]:
melb_df[categorical_vars]

Suburb Type Method        SellerG  CouncilArea  \
0        Abbotsford    h      S         Biggin        Yarra   
1        Abbotsford    h      S         Biggin        Yarra   
2        Abbotsford    h     SP         Biggin        Yarra   
3        Abbotsford    h     PI         Biggin        Yarra   
4        Abbotsford    h     VB         Nelson        Yarra   
...             ...  ...    ...            ...          ...   
10691      Werribee    h      S  hockingstuart      Wyndham   
10692  Williamstown    h     SP       Williams  Hobsons Bay   
10693  Williamstown    h      S          Raine  Hobsons Bay   
10694  Williamstown    h     PI        Sweeney  Hobsons Bay   
10695    Yarraville    h     SP        Village  Maribyrnong   

                  Regionname  
0      Northern Metropolitan  
1      Northern Metropolitan  
2      Northern Metropolitan  
3      Northern Metropolitan  
4      Northern Metropolitan  
...                      ...  
10691   Western Metropolitan  
10692   Western Metropolitan  
10693   Western Metropolitan  
10694   Western Metropolitan  
10695   Western Metropolitan  

[10696 rows x 6 columns]

# Exercises

## Exercise 1 - Queries and use of Pandas.

- Calculate the 5 highest-priced (column `Price`) suburbs on average in Melbourne.

- Calculate the mean, median, max, min, and quartiles of grouping by `Type`
and `SellerG`.

- Retrieve the number of properties in Melbourne that have less or equal than one bathroom and
which are located in suburbs with more than 200 properties, (don't use
`Propertycount`)

## Exercise 2 - OHE and PCA.

- Encode all the categorical columns `Suburb`, `Type`, `Method`, `SellerG`, `CouncilArea`, and `Regionname` of `melb_df` using one hot encoder.

- Standarize numerical columns (Difference each random variable by the mean and divide by their standard deviation) `Bedroom2`, `BuildingArea`, `Longtitude`, `Car`, `Postcode`, `Bathroom`, `Distance`, `Lattitude`, `Rooms`, `suburb_count`, `Price`, `Propertycount`, and `Landsize`.

- Apply PCA to the standarized numerical columns.

- Concat the encodings of PCA and the one hot encoded columns in one dataframe.

## Solution 1 - Queries and use of Pandas.

Calculate the 5 highest-priced (column `Price`) suburbs on average in Melbourne.

In [19]:
(
    melb_df[["Suburb", "Price"]]
    .groupby("Suburb")
    .mean()
    .round(3)
    .sort_values(by="Price", ascending=False)
    .head(5)
)

Price
Suburb                   
Albert Park   1725059.322
Balwyn North  1676786.709
Kew East      1666375.000
Ashburton     1658096.491
Balwyn        1537604.651

Calculate the mean, median, max, min, and quartiles of grouping by `Type`
and `SellerG`.

In [20]:
(
    melb_df[["SellerG", "Type", "Price"]]
    .groupby(["SellerG", "Type"])
    .describe()
)

Price                                          \
                             count          mean            std        min   
SellerG                 Type                                                 
@Realty                 h      2.0  6.120000e+05  159806.132548   499000.0   
ASL                     h      3.0  1.393333e+06  433628.104870  1090000.0   
Abercromby's            h     10.0  1.627100e+06  473150.070156   860000.0   
                        t      4.0  1.820750e+06  619135.620577  1253000.0   
                        u      3.0  1.060000e+06  725809.892465   500000.0   
...                            ...           ...            ...        ...   
hockingstuart/Advantage u      1.0  3.300000e+05            NaN   330000.0   
hockingstuart/Barry     h      1.0  1.265000e+06            NaN  1265000.0   
hockingstuart/Village   u      1.0  3.450000e+05            NaN   345000.0   
iOne                    t      1.0  6.200000e+05            NaN   620000.0   
iTRAK                   u      1.0  4.950000e+05            NaN   495000.0   

                                                                          
                                    25%        50%        75%        max  
SellerG                 Type                                              
@Realty                 h      555500.0   612000.0   668500.0   725000.0  
ASL                     h     1145000.0  1200000.0  1545000.0  1890000.0  
Abercromby's            h     1395000.0  1632500.0  1870750.0  2500000.0  
                        t     1340750.0  1735000.0  2215000.0  2560000.0  
                        u      650000.0   800000.0  1340000.0  1880000.0  
...                                 ...        ...        ...        ...  
hockingstuart/Advantage u      330000.0   330000.0   330000.0   330000.0  
hockingstuart/Barry     h     1265000.0  1265000.0  1265000.0  1265000.0  
hockingstuart/Village   u      345000.0   345000.0   345000.0   345000.0  
iOne                    t      620000.0   620000.0   620000.0   620000.0  
iTRAK                   u      495000.0   495000.0   495000.0   495000.0  

[403 rows x 8 columns]

Retrieve the number of properties in Melbourne that have less or equal than one bathroom and
which are located in suburbs with more than 200 properties, (don't use
`Propertycount`)

In [21]:
property_count = (
    melb_df[["Suburb", "Address"]]
    .groupby("Suburb")
    .count()
    .reset_index()
)
property_count

Suburb  Address
0        Abbotsford       56
1        Aberfeldie       42
2      Airport West       67
3       Albert Park       59
4            Albion       41
..              ...      ...
104        Werribee       50
105  West Footscray      105
106    Williamstown      103
107         Windsor       60
108      Yarraville      164

[109 rows x 2 columns]

In [22]:
populated_suburbs = property_count[property_count["Address"] > 200]
melb_df[
    (melb_df["Suburb"].isin(populated_suburbs["Suburb"])) &
    (melb_df["Bathroom"] <= 1)
]

Unnamed: 0          Suburb          Address  Rooms Type      Price  \
704           704  Bentleigh East     18 Mervin St      2    h  1410000.0   
705           705  Bentleigh East  1/69 Purtell St      3    t   760000.0   
707           707  Bentleigh East     2a Bessie St      2    u   746000.0   
708           708  Bentleigh East    14a Daphne St      1    u   580000.0   
710           710  Bentleigh East   93 Parkmore Rd      3    h  1206000.0   
...           ...             ...              ...    ...  ...        ...   
10667       10667         Preston     61 Dundas St      3    h   910000.0   
10668       10668       Reservoir      1E Black St      3    t   640000.0   
10671       10671       Reservoir      4 Newton St      3    h   800000.0   
10672       10672       Reservoir       92 Pine St      2    u   625500.0   
10673       10673        Richmond    170 Coppin St      4    h  1800000.0   

      Method        SellerG        Date  Distance  Postcode  Bedroom2  \
704        S  hockingstuart   3/09/2016      13.9    3165.0       2.0   
705       PI  hockingstuart   3/09/2016      13.9    3165.0       3.0   
707        S  hockingstuart   3/12/2016      13.9    3165.0       2.0   
708        S       Woodards   3/12/2016      13.9    3165.0       1.0   
710        S            C21   3/12/2016      13.9    3165.0       3.0   
...      ...            ...         ...       ...       ...       ...   
10667     SP         Nelson  26/08/2017       8.4    3072.0       3.0   
10668     PI            Ray  26/08/2017      12.0    3073.0       3.0   
10671     PI            Ray  26/08/2017      12.0    3073.0       3.0   
10672      S             RW  26/08/2017      12.0    3073.0       2.0   
10673      S         Biggin  26/08/2017       2.4    3121.0       4.0   

       Bathroom  Car  Landsize  BuildingArea CouncilArea  Lattitude  \
704         1.0  2.0     768.0    137.374429   Glen Eira  -37.93050   
705         1.0  2.0     287.0    137.374429   Glen Eira  -37.92610   
707         1.0  2.0     281.0     95.000000   Glen Eira  -37.92550   
708         1.0  1.0     105.0    137.374429   Glen Eira  -37.91810   
710         1.0  2.0     603.0    137.374429   Glen Eira  -37.91550   
...         ...  ...       ...           ...         ...        ...   
10667       1.0  2.0     276.0     96.000000     Darebin  -37.75277   
10668       1.0  1.0     224.0    128.000000     Darebin  -37.71289   
10671       1.0  2.0     585.0     90.000000     Darebin  -37.73173   
10672       1.0  2.0     343.0    111.000000     Darebin  -37.71626   
10673       1.0  4.0     418.0    137.374429       Yarra  -37.82368   

       Longtitude             Regionname  Propertycount  suburb_count  \
704     145.05470  Southern Metropolitan        10969.0           249   
705     145.05660  Southern Metropolitan        10969.0           249   
707     145.07960  Southern Metropolitan        10969.0           249   
708     145.05860  Southern Metropolitan        10969.0           249   
710     145.07230  Southern Metropolitan        10969.0           249   
...           ...                    ...            ...           ...   
10667   145.00734  Northern Metropolitan        14577.0           239   
10668   145.01967  Northern Metropolitan        21650.0           359   
10671   145.02884  Northern Metropolitan        21650.0           359   
10672   144.99866  Northern Metropolitan        21650.0           359   
10673   145.00271  Northern Metropolitan        14949.0           255   

      CouncilArea_mode  
704          Glen Eira  
705          Glen Eira  
707          Glen Eira  
708          Glen Eira  
710          Glen Eira  
...                ...  
10667          Darebin  
10668          Darebin  
10671          Darebin  
10672          Darebin  
10673            Yarra  

[1018 rows x 23 columns]

## Solution: Using One Hot Encoder

We first import from scikit-learn the class One Hot Encoder, which implements
the codification algorithm. Then we make an instance of this class.

In [23]:
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder()

The categorical variables we have to encode are `Suburb`, `Type`, `Method`,
`SellerG`, `CouncilArea`, and `Regionname`. Each variables contains the
following number of unique categories.

In [24]:
nof_categ = 0
for categ in categorical_vars:
    print(f"Variable {categ} has {melb_df[categ].nunique()} categories")
    nof_categ += melb_df[categ].nunique()
print(f"Total number of categories {nof_categ}")


Variable Suburb has 109 categories
Variable Type has 3 categories
Variable Method has 5 categories
Variable SellerG has 219 categories
Variable CouncilArea has 22 categories
Variable Regionname has 4 categories
Total number of categories 362



So we have in total 362 categories counting all the variables. That indicates
that after applying one hot encoding, we should expect 362 columns for each row.

We proceed to fit the algorithm (which will save necessary states variables that
depend on our sample)

In [25]:
melb_df[categorical_vars].shape

(10696, 6)

In [26]:
ohe.fit(melb_df[categorical_vars])

OneHotEncoder()

Once the algorithm is fit, we can transform our original data with shape (10696,
6) to (10696, 362) as we expect after counting all the possible categories.

In [27]:
melb_df_enc = ohe.transform(melb_df[categorical_vars])

Note that melb_df_enc was before a pandas dataframe and now is a compressed
numpy sparse matrix. As the resultant matrix contains very few 1's and multiple
0's, the One Hot algorithm makes this optimization for us. If we want to
retrieve the uncompressed matrix, we can use the method `todense()`.

In [28]:
melb_df_enc.todense()

matrix([[1., 0., 0., ..., 1., 0., 0.],
        [1., 0., 0., ..., 1., 0., 0.],
        [1., 0., 0., ..., 1., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 1.],
        [0., 0., 0., ..., 0., 0., 1.],
        [0., 0., 0., ..., 0., 0., 1.]])

In [29]:
melb_df_enc.todense().shape

(10696, 362)

## Solution: PCA

Let's now use PCA on the numerical data of our dataset. Similarly to One Hot
Encoder, scikit-learn has an implementation of the PCA algorithm utilizing a
class. So we start by making an instance of it.

In [30]:
melb_df[numerical_vars]

Car  Distance  Landsize  BuildingArea  Bedroom2  Bathroom        Date  \
0      1.0       2.5     202.0    137.374429       2.0       1.0   3/12/2016   
1      0.0       2.5     156.0     79.000000       2.0       1.0   4/02/2016   
2      0.0       2.5     134.0    150.000000       3.0       2.0   4/03/2017   
3      1.0       2.5      94.0    137.374429       3.0       2.0   4/03/2017   
4      2.0       2.5     120.0    142.000000       3.0       1.0   4/06/2016   
...    ...       ...       ...           ...       ...       ...         ...   
10691  1.0      14.7     662.0    172.000000       4.0       2.0  26/08/2017   
10692  2.0       6.8     333.0    133.000000       3.0       2.0  26/08/2017   
10693  4.0       6.8     436.0    137.374429       3.0       2.0  26/08/2017   
10694  5.0       6.8     866.0    157.000000       4.0       1.0  26/08/2017   
10695  1.0       6.3     362.0    112.000000       4.0       1.0  26/08/2017   

       Rooms      Price  Postcode  Lattitude  suburb_count  Propertycount  \
0          2  1480000.0    3067.0  -37.79960            56         4019.0   
1          2  1035000.0    3067.0  -37.80790            56         4019.0   
2          3  1465000.0    3067.0  -37.80930            56         4019.0   
3          3   850000.0    3067.0  -37.79690            56         4019.0   
4          4  1600000.0    3067.0  -37.80720            56         4019.0   
...      ...        ...       ...        ...           ...            ...   
10691      4   635000.0    3030.0  -37.89327            50        16166.0   
10692      3  1031000.0    3016.0  -37.85927           103         6380.0   
10693      3  1170000.0    3016.0  -37.85274           103         6380.0   
10694      4  2500000.0    3016.0  -37.85908           103         6380.0   
10695      4  1285000.0    3013.0  -37.81188           164         6543.0   

      CouncilArea_mode  Unnamed: 0           Address  Longtitude  
0                Yarra           0      85 Turner St   144.99840  
1                Yarra           1   25 Bloomburg St   144.99340  
2                Yarra           2      5 Charles St   144.99440  
3                Yarra           3  40 Federation La   144.99690  
4                Yarra           4       55a Park St   144.99410  
...                ...         ...               ...         ...  
10691          Wyndham       10691       5 Nuragi Ct   144.64789  
10692      Hobsons Bay       10692     77 Merrett Dr   144.87904  
10693      Hobsons Bay       10693       83 Power St   144.88738  
10694      Hobsons Bay       10694      96 Verdon St   144.89299  
10695      Maribyrnong       10695        6 Agnes St   144.88449  

[10696 rows x 17 columns]

When training the `PCA` class instance, scikit-learn takes care of centering the
data by subtracting the mean. However, it is also recommended to *standardize*
or at least *scale* the original array to ensure that all variables are in the
same units.


In [31]:
melb_df_num = (
    (melb_df[numerical_vars] - melb_df[numerical_vars].mean(axis=0)
) / melb_df[numerical_vars].std(axis=0))

/tmp/ipykernel_206517/3769171654.py:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  (melb_df[numerical_vars] - melb_df[numerical_vars].mean(axis=0)
/tmp/ipykernel_206517/3769171654.py:3: FutureWarning: The default value of numeric_only in DataFrame.std is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  ) / melb_df[numerical_vars].std(axis=0))


In [32]:
melb_df_num

Address  Bathroom  Bedroom2  BuildingArea       Car CouncilArea_mode  \
0         NaN -0.726214 -0.867214      0.000000 -0.565460              NaN   
1         NaN -0.726214 -0.867214     -0.598739 -1.634229              NaN   
2         NaN  0.817730  0.209730      0.129499 -1.634229              NaN   
3         NaN  0.817730  0.209730      0.000000 -0.565460              NaN   
4         NaN -0.726214  0.209730      0.047444  0.503309              NaN   
...       ...       ...       ...           ...       ...              ...   
10691     NaN  0.817730  1.286675      0.355150 -0.565460              NaN   
10692     NaN  0.817730  0.209730     -0.044868  0.503309              NaN   
10693     NaN  0.817730  0.209730      0.000000  2.640847              NaN   
10694     NaN -0.726214  1.286675      0.201297  3.709616              NaN   
10695     NaN -0.726214  1.286675     -0.260262 -0.565460              NaN   

      Date  Distance  Landsize  Lattitude  Longtitude  Postcode     Price  \
0      NaN -1.603407 -0.222499   0.082343    0.155207 -0.452154  0.884437   
1      NaN -1.603407 -0.258758  -0.043895    0.092281 -0.452154  0.000278   
2      NaN -1.603407 -0.276100  -0.065188    0.104866 -0.452154  0.854634   
3      NaN -1.603407 -0.307629   0.123408    0.136329 -0.452154 -0.367294   
4      NaN -1.603407 -0.287135  -0.033248    0.101091 -0.452154  1.122862   
...    ...       ...       ...        ...         ...       ...       ...   
10691  NaN  1.558321  0.140093  -1.342315   -4.256053 -1.078479 -0.794473   
10692  NaN -0.489028 -0.119239  -0.825198   -1.346970 -1.315466 -0.007670   
10693  NaN -0.489028 -0.038050  -0.725881   -1.242009 -1.315466  0.268506   
10694  NaN -0.489028  0.300895  -0.822308   -1.171405 -1.315466  2.911050   
10695  NaN -0.618607 -0.096380  -0.104428   -1.278380 -1.366249  0.496996   

       Propertycount     Rooms  Unnamed: 0  suburb_count  
0          -0.924454 -0.892052   -1.731808     -1.042652  
1          -0.924454 -0.892052   -1.731484     -1.042652  
2          -0.924454  0.181461   -1.731160     -1.042652  
3          -0.924454  0.181461   -1.730836     -1.042652  
4          -0.924454  1.254975   -1.730513     -1.042652  
...              ...       ...         ...           ...  
10691       1.871696  1.254975    1.730513     -1.126244  
10692      -0.380969  0.181461    1.730836     -0.387848  
10693      -0.380969  0.181461    1.731160     -0.387848  
10694      -0.380969  1.254975    1.731484     -0.387848  
10695      -0.343447  1.254975    1.731808      0.462004  

[10696 rows x 17 columns]

Now we are going to discover how many principal components we are using. We can
start with a reasonable but possibly a bit excessive number of components, like
8 (knowing that we have 13 columns).

In [33]:
from sklearn.decomposition import PCA
pca = PCA(n_components=8)

In [34]:
pca.fit(melb_df_num)
proyected_features = pca.transform(melb_df_num)

ValueError: Input X contains NaN.
PCA does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

In [ ]:
proyected_features.shape

(10696, 8)

Then, we can plot how much information provides each component through a
lineplot.

In [35]:
seaborn.lineplot(x=range(pca.n_components),
                 y=pca.explained_variance_ratio_)
plt.ylabel("Ratio of variance explained")
plt.xlabel("Components")

AttributeError: 'PCA' object has no attribute 'explained_variance_ratio_'

Here we can see that after 4 components, the following ones provide less than
10\% of information. Therefore we can choose between 4 or 5 components.

In [36]:
from sklearn.decomposition import PCA
pca = PCA(n_components=4)
pca.fit(melb_df_num)
proyected_features = pca.transform(melb_df_num)

ValueError: Input X contains NaN.
PCA does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values

In [37]:
sum(pca.explained_variance_ratio_)

AttributeError: 'PCA' object has no attribute 'explained_variance_ratio_'

## Combining numerical and categorical variables

Additionally, we can combine the features obtained from the PCA and the OHE
algorithms. We ended up with two numpy matrices of the following shapes:

`numpy` is a python library that offers comprehensive mathematical functions,
random number generators, linear algebra routines, and a lot more! That's why
most of the machine learning libraries depend on `numpy` as the defacto
functionality to develop algorithms.

In [38]:
proyected_features.shape

NameError: name 'proyected_features' is not defined

In [39]:
melb_df_enc.todense().shape

(10696, 362)


We can combine these two matrices using stack operations. You can stack two or
more matrices horizontally (`hstack`) or vertically (`vstack`). As we want to
combine categorical and numerical features while keeping the same number of
rows, we need to use `hstack`.

In [40]:
import numpy as np
np.hstack([proyected_features, melb_df_enc.todense()]).shape

NameError: name 'proyected_features' is not defined

Finally, we can create a pandas dataframe, adding column names to the final
matrix. The instances of OHE and PCA contain the name of the encoded categories
and the number of components used.

In [41]:
features = np.hstack([proyected_features, melb_df_enc.todense()])
categ_ohe = np.concatenate(ohe.categories_).tolist()
categ_pca = [f"pca_{comp}" for comp in range(pca.n_components_)]

NameError: name 'proyected_features' is not defined

In [42]:
import pandas as pd
merged_features = pd.DataFrame(features, columns=categ_pca + categ_ohe)

NameError: name 'features' is not defined

In [43]:
merged_features

NameError: name 'merged_features' is not defined